src/tests/util/test_database.rs

```diff
@@ -0,0 +1,158 @@
+use crate::env;
+use diesel::prelude::*;
+use diesel::r2d2::{ConnectionManager, Pool, PooledConnection};
+use diesel::sql_query;
+use diesel_migrations::{FileBasedMigrations, MigrationHarness};
+use once_cell::sync::Lazy;
+use rand::Rng;
+use tracing::instrument;
+use url::Url;
+
+struct TemplateDatabase {
+    base_url: Url,
+    pool: Pool<ConnectionManager<PgConnection>>,
+    template_name: String,
+    prefix: String,
+}
+
+impl TemplateDatabase {
+    #[instrument]
+    pub fn instance() -> &'static Self {
+        static INSTANCE: Lazy<TemplateDatabase> = Lazy::new(TemplateDatabase::new);
+        &INSTANCE
+    }
+
+    #[instrument]
+    fn new() -> Self {
+        let base_url = env("TEST_DATABASE_URL");
+        let base_url = Url::parse(&base_url).expect("failed to parse TEST_DATABASE_URL");
+
+        let prefix = base_url.path().strip_prefix('/');
+        let prefix = prefix.expect("failed to parse database name").to_string();
+
+        // Having only a single database management connection was causing
+        // contention, so this is using a connection pool to reduce unnecessary
+        // waiting times for the tests.
+        let pool = Pool::builder()
+            .max_size(10)
+            .min_idle(Some(0))
+            .build_unchecked(ConnectionManager::new(base_url.as_ref()));
+
+        // Get a connection from the pool, and create the template database
+        let mut conn = pool.get().expect("failed to connect to the database");
+
+        let template_name = format!("{prefix}_template");
+        let _ = drop_database(&template_name, &mut conn);
+        create_template_database(&template_name, &mut conn)
+            .expect("failed to create template database");
```

Comment:
I'm wondering if it would be better to have a random per-test-run prefix to allow for multiple test runs in parallel? Since we are dropping the template database on `Drop` anyway, I don't think this adds much risk of polluting the user's database instance.

Reply:
✅